### PR TITLE
Add some event tests

### PR DESCRIPTION
**Pull request by feuGeneA**
These tests were created in the course of investigating #588 .  That investigation has not been completed, but these tests seem good enough to merge already.

### CODE REVIEW

File: packages/hardhat-core/src/internal/core/providers/backwards-compatibility.ts [hunk 1]
```diff
@@ -9,6 +9,13 @@ import {
 } from "../../../types";
 import { EventEmitterWrapper } from "../../util/event-emitter";
 
+/**
+ * Hardhat predates the EIP1193 (Javascript Ethereum Provider) standard. It was
+ * built following a draft of that spec, but then it changed completely. We
+ * still need to support the draft api, but internally we use EIP1193. So we
```
[the remark this answers is not in the document]
Comment (feuGeneA) [reply packages/hardhat-core/src/internal/core/providers/backwards-compatibility.ts:15]: Addressed in https://github.com/nomiclabs/hardhat/pull/1792/commits/2e4fdaf3aba321e90dd9041ea4390f22339b4c78

File: packages/hardhat-ethers/src/internal/ethers-websocket-provider-proxy.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import { createUpdatableTargetProxy } from "./updatable-target-proxy";
+
+/**
+ * This method returns a proxy that uses an underlying provider for everything.
+ *
+ * This underlying provider is replaced by a new one after a successful hardhat_reset,
+ * because ethers providers can have internal state that returns wrong results after
+ * the network is reset.
+ */
+export function createProviderProxy(
```
Comment (alcuadrado) [inline packages/hardhat-ethers/src/internal/ethers-websocket-provider-proxy.ts:18]: Is this only meant for testing?

File: packages/hardhat-ethers/src/internal/websocket.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import {
+  getContractAt,
+  getContractFactory,
+  getSigner,
+  getSigners,
+} from "./helpers";
+import type * as ProviderProxyT from "./ethers-websocket-provider-proxy";
+import "./type-extensions";
+
+extendEnvironment((hre) => {
```
Comment (alcuadrado) [inline packages/hardhat-ethers/src/internal/websocket.ts:14]: Is this meant to be an alternative version of the plugin?

File: packages/hardhat-ethers/test/fixture-projects/hardhat-project-websocket/artifacts/contracts/AmbiguousLibrary.sol/AmbiguousLibrary.dbg.json [hunk 4]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (alcuadrado) [inline packages/hardhat-ethers/test/fixture-projects/hardhat-project-websocket/artifacts/contracts/AmbiguousLibrary.sol/AmbiguousLibrary.dbg.json:1]: I guess this `artifacts/` folder should be added to a `.gitignore`

File: packages/hardhat-ethers/src/internal/ethers-websocket-provider-proxy.ts (hunk 2 above)
Comment (feuGeneA) [reply packages/hardhat-ethers/src/internal/ethers-websocket-provider-proxy.ts:18]: Sorry, I should have been more clear here.  This change was only made in support of the websocket+ganache test.  To be honest I hacked my way through that test to make it work.  Maybe it's actually not as ready for "merge" as the rest of the tests in here.

File: packages/hardhat-ethers/test/fixture-projects/hardhat-project-websocket/artifacts/contracts/AmbiguousLibrary.sol/AmbiguousLibrary.dbg.json (hunk 4 above)
Comment (feuGeneA) [reply packages/hardhat-ethers/test/fixture-projects/hardhat-project-websocket/artifacts/contracts/AmbiguousLibrary.sol/AmbiguousLibrary.dbg.json]: Addressed in https://github.com/nomiclabs/hardhat/pull/1792/commits/165a656d9eb7832d8ce4048a77205e24eb83965b and  https://github.com/nomiclabs/hardhat/pull/1792/commits/4f922c9e01aa54d67e4aded0b7828b3f58be72ec

File: packages/hardhat-ethers/src/internal/websocket.ts (hunk 3 above)
Comment (feuGeneA) [reply packages/hardhat-ethers/src/internal/websocket.ts:14]: No? haha. I just hacked my way through getting the test to work.  I'm not even sure exactly what it is I did. :sweat_smile:

File: packages/hardhat-ethers/src/internal/ethers-websocket-provider-proxy.ts (hunk 2 above)
Comment (feuGeneA) [reply packages/hardhat-ethers/src/internal/ethers-websocket-provider-proxy.ts]: Eliminated this in https://github.com/nomiclabs/hardhat/pull/1792/commits/3ec7a887d380aaaa1e7e119ad15dcc96d94d3a82

File: packages/hardhat-ethers/src/internal/websocket.ts (hunk 3 above)
Comment (feuGeneA) [reply packages/hardhat-ethers/src/internal/websocket.ts]: Eliminated this in https://github.com/nomiclabs/hardhat/pull/1792/commits/3ec7a887d380aaaa1e7e119ad15dcc96d94d3a82

**Review by alcuadrado — APPROVED** [review]
The current version is so much simpler :)

**Pull request merged**